okhttp/src/main/java/okhttp3/Authenticator.java

```diff
@@ -28,24 +28,30 @@
  * authenticate. If so it is likely that further attempts will not be useful and the authenticator
  * should give up.
  *
+ * <p></p>
+ *
  * <p>When authentication is requested by an origin server, the response code is 401 and the
  * implementation should respond with a new request that sets the "Authorization" header.
  * <pre>   {@code
  *
- *    if (response.request().header("Authorization") != null) {
- *      return null; // Give up, we've already failed to authenticate.
- *    }
+ *  Request authenticate(Route route, Response response) {
+ *      if (response.request().header("Authorization") != null) {
+ *        return null; // Give up, we've already failed to authenticate.
+ *      }
+ *
+ *      String credential = Credentials.basic(...)
+ *        return response.request().newBuilder()
+ *          .header("Authorization", credential)
+ *          .build();
+ *  }
  *
- *    String credential = Credentials.basic(...)
- *    return response.request().newBuilder()
- *        .header("Authorization", credential)
- *        .build();
  * }</pre>
  *
  * <p>When authentication is requested by a proxy server, the response code is 407 and the
  * implementation should respond with a new request that sets the "Proxy-Authorization" header.
  * <pre>   {@code
  *
+ *  Request authenticate(Route route, Response response) {
```

Comment:
My reading of https://github.com/square/okhttp/issues/2435 was this would be not part of the interface, but a subclass that could be checked for.  This seems like a pretty big API breaking change as is.

Reply:
Yes, You are right. You mean to add a method to avoid breaking change? It will be nice if the iusses can be solved in 3.10.0. etc. #2514 #3429 looks like familiar.

Reply:
If we prefer interfaces to an abstract class, then something like...

```
public interface PreemptiveAuthenticator extends Authenticator  {    
   /**
    * .... Returns null if no authentication.
    */
   @Nullable Request authenticate(Request request) throws IOException;
}
```